promql/test.go

```diff
@@ -445,59 +508,66 @@ func (t *Test) exec(tc testCommand) error {
 		}
 
 	case *evalCmd:
-		q, err := t.QueryEngine().NewInstantQuery(t.storage, cmd.expr, cmd.start)
+		queries, err := atModifierTestCases(cmd.expr, cmd.start)
 		if err != nil {
 			return err
 		}
-		defer q.Close()
-		res := q.Exec(t.context)
-		if res.Err != nil {
-			if cmd.fail {
-				return nil
+		queries = append([]atModifierTestCase{{expr: cmd.expr, evalTime: cmd.start}}, queries...)
+		for _, iq := range queries {
+			q, err := t.QueryEngine().NewInstantQuery(t.storage, iq.expr, iq.evalTime)
+			if err != nil {
+				return err
+			}
+			defer q.Close()
+			res := q.Exec(t.context)
+			if res.Err != nil {
+				if cmd.fail {
+					return nil
+				}
+				return errors.Wrapf(res.Err, "error evaluating query %q (line %d)", iq.expr, cmd.line)
+			}
+			if res.Err == nil && cmd.fail {
+				return errors.Errorf("expected error evaluating query %q (line %d) but got none", iq.expr, cmd.line)
+			}
+			err = cmd.compareResult(res.Value)
+			if err != nil {
+				return errors.Wrapf(err, "error in %s %s", cmd, iq.expr)
 			}
-			return errors.Wrapf(res.Err, "error evaluating query %q (line %d)", cmd.expr, cmd.line)
-		}
-		if res.Err == nil && cmd.fail {
-			return errors.Errorf("expected error evaluating query %q (line %d) but got none", cmd.expr, cmd.line)
-		}
-
-		err = cmd.compareResult(res.Value)
-		if err != nil {
-			return errors.Wrapf(err, "error in %s %s", cmd, cmd.expr)
-		}
 
-		// Check query returns same result in range mode,
-		// by checking against the middle step.
-		q, err = t.queryEngine.NewRangeQuery(t.storage, cmd.expr, cmd.start.Add(-time.Minute), cmd.start.Add(time.Minute), time.Minute)
-		if err != nil {
-			return err
-		}
-		rangeRes := q.Exec(t.context)
-		if rangeRes.Err != nil {
-			return errors.Wrapf(rangeRes.Err, "error evaluating query %q (line %d) in range mode", cmd.expr, cmd.line)
-		}
-		defer q.Close()
-		if cmd.ordered {
-			// Ordering isn't defined for range queries.
-			return nil
-		}
-		mat := rangeRes.Value.(Matrix)
-		vec := make(Vector, 0, len(mat))
-		for _, series := range mat {
-			for _, point := range series.Points {
-				if point.T == timeMilliseconds(cmd.start) {
-					vec = append(vec, Sample{Metric: series.Metric, Point: point})
-					break
+			// Check query returns same result in range mode,
```

Comment:
You mean this should be separate unit test in itself?

Reply:
Or a separate thing at a higher level and not clubbed here?

Reply:
No, i mean that it should be a single extra thing we test - like how the range query is done.

Reply:
We want this @ timestamp modification to be run on both instant queries and range queries. Are you suggesting we duplicate the code here?

Reply:
Why do you think we need it on both?

Reply:
Same reason as to why we need other queries for both. At API level query_range may be a syntactic sugar, but the engine does not execute it in similar manner. Additionally, range query will test proper duplication of results.

Reply:
I guess I am gonna keep it on both, already found something unusual with `timestamp(timestamp(metric @ x))` on range query

Reply:
> timestamp(timestamp(metric @ x))

I had missed including `timestamp` in non step invariant functions

Reply:
Uhm, timestamp is step invariant. It doesn't care what the evaluation time is.

Reply:
`timestamp(timestamp(...))` is not step invariant because the timestamp of the result of the inner `timestamp(..)` depends on the eval time, hence affects the outer `timestamp(...)`. So while `timestamp(..)` itself is step invariant, nested timestamp() is not.

For that matter, I think `timestamp(...)` with anything not vector selector will not be step invariant.

Reply:
So I guess it is safe to assume timestamp() is not step invariant, or have a special case for vector selector inside it

Reply:
> So while timestamp(..) itself is step invariant, nested timestamp() is not.

You're confusing a function being step invariant with an expression being step invariant. The timestamp function is step invariant, it never touches the evaluation time as it only looks at the samples being passed in to it.

> For that matter, I think timestamp(...) with anything not vector selector will not be step invariant.

No, timestamp() is always step invariant. If you're seeing otherwise, that's a bug in the engine somewhere as it means  samples in instant vectors have the wrong timestamp.

Reply:
I will have a special case for timestamp() in wrapping then

Reply:
First you should ask why the wrong timestamp is getting that far, to see if there's other issues hiding.